<a href="https://colab.research.google.com/github/MoonByeongPil/Test/blob/main/HDNET_TensorFlow%EC%97%B0%EC%8A%B5%20%EB%B0%8F%20%ED%99%95%EC%9D%B8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##### Copyright 2018 The TensorFlow Authors.

Licensed under the Apache License, Version 2.0 (the "License");

In [ ]:
#@title Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# 텐서플로 2.0 시작하기: 초보자용

<table class="tfo-notebook-buttons" align="left">
  <td>
    <a target="_blank" href="https://www.tensorflow.org/tutorials/quickstart/beginner"><img src="https://www.tensorflow.org/images/tf_logo_32px.png" />TensorFlow.org에서 보기</a>
  </td>
  <td>
    <a target="_blank" href="https://colab.research.google.com/github/tensorflow/docs-l10n/blob/master/site/ko/tutorials/quickstart/beginner.ipynb"><img src="https://www.tensorflow.org/images/colab_logo_32px.png" />구글 코랩(Colab)에서 실행하기</a>
  </td>
  <td>
    <a target="_blank" href="https://github.com/tensorflow/docs-l10n/blob/master/site/ko/tutorials/quickstart/beginner.ipynb"><img src="https://www.tensorflow.org/images/GitHub-Mark-32px.png" />깃허브(GitHub) 소스 보기</a>
  </td>
  <td>
    <a href="https://storage.googleapis.com/tensorflow_docs/docs-l10n/site/ko/tutorials/quickstart/beginner.ipynb"><img src="https://www.tensorflow.org/images/download_logo_32px.png" />Download notebook</a>
  </td>
</table>

Note: 이 문서는 텐서플로 커뮤니티에서 번역했습니다. 커뮤니티 번역 활동의 특성상 정확한 번역과 최신 내용을 반영하기 위해 노력함에도
불구하고 [공식 영문 문서](https://www.tensorflow.org/?hl=en)의 내용과 일치하지 않을 수 있습니다.
이 번역에 개선할 부분이 있다면
[tensorflow/docs-l10n](https://github.com/tensorflow/docs-l10n/) 깃헙 저장소로 풀 리퀘스트를 보내주시기 바랍니다.
문서 번역이나 리뷰에 참여하려면
[docs-ko@tensorflow.org](https://groups.google.com/a/tensorflow.org/forum/#!forum/docs-ko)로
메일을 보내주시기 바랍니다.

이 문서는 [구글 코랩](https://colab.research.google.com/notebooks/welcome.ipynb)(Colaboratory) 노트북 파일입니다. 파이썬 프로그램을 브라우저에서 직접 실행할 수 있기 때문에 텐서플로를 배우고 사용하기 좋은 도구입니다:

1. 파이썬 런타임(runtime)에 연결하세요: 메뉴 막대의 오른쪽 상단에서 *CONNECT*를 선택하세요.
2. 노트북의 모든 코드 셀(cell)을 실행하세요: *Runtime* > *Run all*을 선택하세요.

더 많은 예제와 자세한 안내는 [텐서플로 튜토리얼](https://www.tensorflow.org/tutorials/)을 참고하세요.

먼저 프로그램에 텐서플로 라이브러리를 임포트합니다:

In [ ]:
import tensorflow as tf

[MNIST 데이터셋](http://yann.lecun.com/exdb/mnist/)을 로드하여 준비합니다. 샘플 값을 정수에서 부동소수로 변환합니다:

In [ ]:
x = tf.constant([[[1, 2, 3], [4, 5, 6]]])
x.numpy()

array([[[1, 2, 3],
        [4, 5, 6]]], dtype=int32)

층을 차례대로 쌓아 `tf.keras.Sequential` 모델을 만듭니다. 훈련에 사용할 옵티마이저(optimizer)와 손실 함수를 선택합니다:

In [ ]:
y = tf.constant([[[7, 8, 9], [10, 11, 12]]])
y.numpy()

array([[[ 7,  8,  9],
        [10, 11, 12]]], dtype=int32)

모델을 훈련하고 평가합니다:

In [ ]:
z=tf.concat([x,y], axis=0)
z.numpy

<bound method _EagerTensorBase.numpy of <tf.Tensor: shape=(2, 2, 3), dtype=int32, numpy=
array([[[ 1,  2,  3],
        [ 4,  5,  6]],

       [[ 7,  8,  9],
        [10, 11, 12]]], dtype=int32)>>

In [ ]:
tf.reduce_sum(z, 2)


<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[ 6, 15],
       [24, 33]], dtype=int32)>

In [ ]:
z

<tf.Tensor: shape=(2, 2, 3), dtype=int32, numpy=
array([[[ 1,  2,  3],
        [ 4,  5,  6]],

       [[ 7,  8,  9],
        [10, 11, 12]]], dtype=int32)>

In [ ]:
tf.square(z)

<tf.Tensor: shape=(2, 2, 3), dtype=int32, numpy=
array([[[  1,   4,   9],
        [ 16,  25,  36]],

       [[ 49,  64,  81],
        [100, 121, 144]]], dtype=int32)>

In [ ]:
k=tf.expand_dims(z,-1)
k

<tf.Tensor: shape=(2, 2, 3, 1), dtype=int32, numpy=
array([[[[ 1],
         [ 2],
         [ 3]],

        [[ 4],
         [ 5],
         [ 6]]],


       [[[ 7],
         [ 8],
         [ 9]],

        [[10],
         [11],
         [12]]]], dtype=int32)>

In [ ]:
tf.reduce_sum(k,3)


<tf.Tensor: shape=(2, 2, 3), dtype=int32, numpy=
array([[[ 1,  2,  3],
        [ 4,  5,  6]],

       [[ 7,  8,  9],
        [10, 11, 12]]], dtype=int32)>

In [ ]:
 z = tf.constant([[[[1,1],[2,2],[3,3]],[[4,4],[5,5],[6,6]]],[[[7,7],[8,8],[9,9]],[[10,10],[11,11],[12,12]]]])
 z

<tf.Tensor: shape=(2, 2, 3, 2), dtype=int32, numpy=
array([[[[ 1,  1],
         [ 2,  2],
         [ 3,  3]],

        [[ 4,  4],
         [ 5,  5],
         [ 6,  6]]],


       [[[ 7,  7],
         [ 8,  8],
         [ 9,  9]],

        [[10, 10],
         [11, 11],
         [12, 12]]]], dtype=int32)>

In [ ]:
a=tf.square(z)
a

<tf.Tensor: shape=(2, 2, 3, 2), dtype=int32, numpy=
array([[[[  1,   1],
         [  4,   4],
         [  9,   9]],

        [[ 16,  16],
         [ 25,  25],
         [ 36,  36]]],


       [[[ 49,  49],
         [ 64,  64],
         [ 81,  81]],

        [[100, 100],
         [121, 121],
         [144, 144]]]], dtype=int32)>

In [ ]:
b=tf.reduce_sum(a,3)
b

<tf.Tensor: shape=(2, 2, 3), dtype=int32, numpy=
array([[[  2,   8,  18],
        [ 32,  50,  72]],

       [[ 98, 128, 162],
        [200, 242, 288]]], dtype=int32)>

In [ ]:
b2=[[[  2.00,   8.00,  18.00],
        [ 32.00,  50.00,  72.00]],

       [[ 98.00, 128.00, 162.00],
        [200.00, 242.00, 288.00]]]
        #정수로 해놓으면 아래에서 sqrt시 정수 결과 안나와서 에러. 미리 실수로 설정.

In [ ]:
c=tf.sqrt(b2)
c

<tf.Tensor: shape=(2, 2, 3), dtype=float32, numpy=
array([[[ 1.4142134,  2.8284268,  4.2426405],
        [ 5.6568537,  7.0710673,  8.485281 ]],

       [[ 9.899494 , 11.313707 , 12.727922 ],
        [14.142136 , 15.556349 , 16.970562 ]]], dtype=float32)>

In [ ]:
d=tf.expand_dims(c,-1)
d


<tf.Tensor: shape=(2, 2, 3, 1), dtype=float32, numpy=
array([[[[ 1.4142134],
         [ 2.8284268],
         [ 4.2426405]],

        [[ 5.6568537],
         [ 7.0710673],
         [ 8.485281 ]]],


       [[[ 9.899494 ],
         [11.313707 ],
         [12.727922 ]],

        [[14.142136 ],
         [15.556349 ],
         [16.970562 ]]]], dtype=float32)>

훈련된 이미지 분류기는 이 데이터셋에서 약 98%의 정확도를 달성합니다. 더 자세한 내용은 [TensorFlow 튜토리얼](https://www.tensorflow.org/tutorials/)을 참고하세요.